lib/vfscore: Auto-mounting scheme with 3 tables

skuenzer opened the pull request:
### Description of changes

This PR introduces a fairly flexible mounting mechanism by
using three filesystem tables:
1) compiled-in, configured via menuconfig
2) user-provided, read entries from kernel parameters ("vfs.fstab")
3) fall-back, similar to "compiled-in"

If auto-mounting is activated, entries of the comiled-in table (1) are mounted first. Afterwards, compiled-in fall-back entries (2) are only processed if the user did not provide any mount entries via the kernel command line (3). The compiled-in tables can be empty and can keep up to four entries.

As part of this PR, The `extract` option of `initrd` replaces the virtual `initrd` filesystem driver. This is done in order to disconnect archive extraction from filesystem mounting and provides greater flexibility for setting up a VFS layout. The traditional initrd mode can be achieved with two entries: First, a RAMFS volume needs to be mounted to `/`. The second entry is used to `extract` virtual fs driver to unpack a CPIO archive to the root volume `/`.


> [!CAUTION]
> This is a breaking change and this PR should be checked in conjunction with [unikraft/kraftkit#1154][0] and merged simultaneously with [#1154][0] to avoid incompatibility issues. Simultaneously, all [catalog](https://github.com/unikraft/catalog) entries should be retriggered against Unikraft's `staging` branch following the merge. Since Unikraft is pre-v1.0, no backwards compatibility is included in this refactor.

[0]: https://github.com/unikraft/unikraft/pull/1154

skuenzer committed:
lib/vfscore: Revert extract option for initrd

This commit reverts the following two commits from PR #1201:
- Commit 4bec77cc0b7d ("lib/vfscore: initrd: "extract" option")
- Commit d1751e01292f ("lib/vfscore: Workaround for initrd extract")

This PR introduces a different auto-mount mechanism using three tables
and makes these two commits obsolete.

Checkpatch-Ignore: GIT_COMMIT_ID
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/vfscore: Refactor virtual "initrd" fs driver towards "extract"

The commit refactors the virtual filesystem "initrd" to "extract".
This is done in order to make the archive extraction step of an initrd
more visible and separate from filesystem mounting. "extract" is
not mounting an underlying ramfs volume. This must be explicitly done
by a fstab entry which provides greater flexibility in setting up
the vfs layout.
At the moment "extract" supports only cpio archives and uses the
source device field to distinguish between an embedded archive and
an initrd module provided by a bootloader.

Checkpatch-Ignore: AVOID_EXTERNS
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukcpio: Treat archive buffer as const

We update the `ukcpio_extract()` function signature so that a given
archive buffer is treated as `const` by the compiler. The extraction
process is a read-only operation of the input buffer.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/vfscore: automount: Improved volume parsing

This commit integrates mainly two improvements:
- Updates to `vfscore_mount_volume()` so that volume entries are treated
  as `const`. The code was already ready for it but it was not propagated.
- Rewrite of `vfscore_parse_volume()` so that it stops relying on
  assertions to fail on incorrect user input via the kernel command
  line (vfs.fstab). Assertions can be compiled out for performance builds
  and, when compiled in, a crash by assertions does not cleanly shut down
  the unikernel. With this commit, the functions do return an according
  error code instead. Any calling code is able to take according actions
  on incorrect user input.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/vfscore: Automount scheme with 3 tables

This commit introduces a fairly flexible mounting mechanism by
using three filesystem tables:
1) compiled-in, configured via menuconfig
2a) user-provided, read entries from kernel parameters ("vfs.fstab")
2b) fall-back, similar to "compiled-in"

If auto-mounting is activated, entries of the comiled-in table (1) are
mounted first. Afterwards, compiled-in fall-back entries (2b) are only
processed if the user did not provide any mount entries via the kernel
command line (2a). The compiled-in tables can be empty.
Note that each of the compiled-in tables are limited to four entries. The
size of the user-provided table can be configured.

Checkpatch-Ignore: LONG_LINE_STRING
Checkpatch-Ignore: LINE_SPACING
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

StefanJum commented:
@skuenzer just to make sure I'm testing it right, selecting `User-provided filesystem table` -> `Fall-back entries` -> `9pfs` in the config and passing no kernel parameters at runtime should be equivalent to the old vfscore automount feature, right?

skuenzer commented:
> @skuenzer just to make sure I'm testing it right, selecting `User-provided filesystem table` -> `Fall-back entries` -> `9pfs` in the config and passing no kernel parameters at runtime should be equivalent to the old vfscore automount feature, right?

The full equivalent is configuring 9pfs in the compiled-in table. Then this will be mounted always, irrespective of what the user provides on the kernel command line. The fall-back table is only taken into account if there is nothing provided by the user on the kernel command line.

nderjung reviewed (approved):
Adding my review, I have tested this and it works as expected for me. Thanks!

Reviewed-by: Alexander Jung <alex@unikraft.io>

StefanJum commented on lib/vfscore/automount.c:
```diff
@@ -50,18 +50,20 @@
-#endif /* CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS */
 
-#if CONFIG_LIBVFSCORE_FSTAB
+#if CONFIG_LIBVFSCORE_AUTOMOUNT_UP
 #define LIBVFSCORE_FSTAB_VOLUME_ARGS_SEP			':'
 #define LIBVFSCORE_FSTAB_UKOPTS_ARGS_SEP			','
```
If we do this, compiling with `Compiled-in filesystem table` and without `User-provided filesystem table` config options will fail with `vfscore/automount.c:466:41: error: LIBVFSCORE_FSTAB_UKOPTS_ARGS_SEP’ undeclared`.
Is there any reason to add these in a guard and not just always define them?

skuenzer replied:
Ouh, good point, there is one tiny mistake then. But you have actually good point, they don't harm if they are unused, no memory will be used. I'll remove this guard. Thanks for spotting this.

StefanJum reviewed (changes requested):
See the one comment, all good besides that.

razvand reviewed (approved):
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

skuenzer committed:
lib/vfscore: Improve ukopt parsing (fix implicit deconst)

This commit improves the parsing of Unikraft mount options. C-strings
are avoided for compile time tables so that only entries provided by the
user need to be parsed. The parsing function is rewritten so that:
1) it makes re-use of the `next_arg()` function
2) it avoids an implicit deconst operation caused by `strchr`
   by parsing ukopts together with the user-provided vfs.fstab entries.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer commented:
> Reviewed-by: Razvan Deaconescu [razvand@unikraft.io](mailto:razvand@unikraft.io) Approved-by: Razvan Deaconescu [razvand@unikraft.io](mailto:razvand@unikraft.io)

Please note that I added another commit on top because I noticed that ukopts where implicitly deconst'd while parsing. There is a bit more described in the commit subject, however, this change has effect of configuring Unikraft mount options, so I assume it affects maybe KraftKit. please have a look @razvand, @nderjung .

StefanJum reviewed (approved):
All good on my side now, thanks.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

nderjung reviewed (approved):
> Please note that I added another commit on top because I noticed that ukopts where implicitly deconst'd while parsing. There is a bit more described in the commit subject, however, this change has effect of configuring Unikraft mount options, so I assume it affects maybe KraftKit. please have a look @razvand, @nderjung .

I have re-tested the changes with [the accompanying changes to KraftKit](https://github.com/unikraft/kraftkit/pull/1154) and everything still works as expected.

Reviewed-by: Alexander Jung <alex@unikraft.io>

razvand commented:
I'm testing it myself before the final approval.

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
